tao_compiler/mlir/disc/transforms/disc_transform_legalize_to_loop.cc

```diff
@@ -0,0 +1,242 @@
+#include "tensorflow/compiler/mlir/disc/transforms/lhlo_elemental_utils.h"
+#include "tensorflow/compiler/mlir/disc/transforms/placement_utils.h"
+
+// This file implements logic to legalize transform fusion pattern to loop.
+
+namespace mlir {
+namespace disc_ral {
+
+namespace {
+
+BlockAndValueMapping buildValueMapping(FusionPattern& fusionPattern,
+                                       func::FuncOp funcOp,
+                                       bool reverse = false) {
+  int argPosition = 0;
+  BlockAndValueMapping mapping;
+  auto addMapping = [&](Value v) {
+    if (reverse)
+      mapping.map(funcOp.getArgument(argPosition++), v);
+    else
+      mapping.map(v, funcOp.getArgument(argPosition++));
+  };
+
+  for (Value v : fusionPattern.getOperands()) addMapping(v);
+  for (Value v : fusionPattern.getInternalResults()) addMapping(v);
+  for (Value v : fusionPattern.getResults()) addMapping(v);
+  return mapping;
+}
+
+struct DiscTransformLegalizeToLoopPass
+    : public DiscTransformLegalizeToLoopPassBase<
+          DiscTransformLegalizeToLoopPass> {
+  explicit DiscTransformLegalizeToLoopPass(bool gpuEnabled,
+                                           const std::string& transformFileName,
+                                           bool enableExpensiveChecks)
+      : DiscTransformLegalizeToLoopPassBase<DiscTransformLegalizeToLoopPass>::
+            DiscTransformLegalizeToLoopPassBase() {
+    this->gpuEnabled_ = gpuEnabled;
+    this->transformFileName_ = transformFileName;
+    this->enableExpensiveChecks_ = enableExpensiveChecks;
+  }
+
+  void runOnOperation() override;
+
+  LogicalResult handleCpuFusionOp(OpBuilder& b, Operation* fusion,
+                                  ShapeAnalysis& shapeAnalysis);
+  // Outlines the fusion op to a standalone module op.
+  LogicalResult outlineFusionOp(lmhlo::FusionOp fusionOp,
+                                FusionPattern& fusionPattern,
+                                OwningOpRef<ModuleOp>& m);
+  // Builds a nested pass pipeline to legalize the outlined fusion op.
+  LogicalResult runTransformPipeline(ModuleOp m);
+  // Inlines the lowered IR into the orignal module.
+  LogicalResult inlineTransformedModule(OpBuilder& b, Operation* fusion,
+                                        FusionPattern& fusionPattern,
+                                        ModuleOp m);
+};
+
+LogicalResult DiscTransformLegalizeToLoopPass::outlineFusionOp(
+    lmhlo::FusionOp fusionOp, FusionPattern& fusionPattern,
+    OwningOpRef<ModuleOp>& m) {
+  Location loc = fusionOp->getLoc();
+  m = ModuleOp::create(loc);
+  auto b = OpBuilder::atBlockBegin(m->getBody());
+
+  SmallVector<Type> inputTypes;
+  SmallVector<Type> outputTypes;
+  for (Value v : fusionPattern.getOperands()) {
+    inputTypes.push_back(v.getType());
+  }
+  for (Value v : fusionPattern.getInternalResults()) {
+    inputTypes.push_back(v.getType());
+  }
+  for (Value v : fusionPattern.getResults()) {
```

Comment:
why results are both `input` and `output`? and why internal result are also `input`?

Reply:
1,  output buffers are allocated by DISC and passed to transformed kernel. Same to the internal outputs.
2,  we also returns these output buffers to avoid generated linage on tensor ops to be DCE since tensor ops have no side effect.